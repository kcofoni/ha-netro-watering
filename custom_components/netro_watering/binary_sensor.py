"""Support for Netro watering system."""
from __future__ import annotations

from dataclasses import dataclass
import datetime
import logging
from typing import Any

from homeassistant.components.binary_sensor import (
    BinarySensorDeviceClass,
    BinarySensorEntity,
    BinarySensorEntityDescription,
)
from homeassistant.config_entries import ConfigEntry
from homeassistant.core import HomeAssistant
from homeassistant.helpers.entity_platform import AddEntitiesCallback
from homeassistant.helpers.update_coordinator import CoordinatorEntity
import homeassistant.util.dt as dt_util

from .const import (
    CONF_DEVICE_TYPE,
    CONTROLLER_DEVICE_TYPE,
    DOMAIN,
    EXTRA_STATE_ATTRIBUTE_SEP_LEFT,
    EXTRA_STATE_ATTRIBUTE_SEP_RIGHT,
)
from .coordinator import NetroControllerUpdateCoordinator

_LOGGER = logging.getLogger(__name__)


@dataclass
class NetroRequiredKeysMixin:
    """Mixin for required keys."""

    netro_name: str


@dataclass
class NetroBinarySensorEntityDescription(
    BinarySensorEntityDescription, NetroRequiredKeysMixin
):
    """Defines Netro entity description."""


NETRO_ZONE_WATERING_DESCRIPTION = NetroBinarySensorEntityDescription(
    key="iswatering",
    name="Is it watering ?",
    device_class=BinarySensorDeviceClass.RUNNING,
    translation_key="iswatering",
    netro_name="watering",
    icon="mdi:watering-can-outline",
)


async def async_setup_entry(
    hass: HomeAssistant,
    config_entry: ConfigEntry,
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Set up entry for a Netro sensors."""
    if config_entry.data[CONF_DEVICE_TYPE] == CONTROLLER_DEVICE_TYPE:
        controller: NetroControllerUpdateCoordinator = hass.data[DOMAIN][
            config_entry.entry_id
        ]
        # add zone binary sensors
        async_add_entities(
            [
                NetroZone(
                    controller,
                    NETRO_ZONE_WATERING_DESCRIPTION,
                    zone_key,
                )
                for zone_key in controller.active_zones
            ]
        )


class NetroZone(
    CoordinatorEntity[NetroControllerUpdateCoordinator], BinarySensorEntity
):
    """A sensor implementation for Netro Zone device."""

    entity_description: NetroBinarySensorEntityDescription
    _attr_has_entity_name = True

    def __init__(
        self,
        coordinator: NetroControllerUpdateCoordinator,
        description: NetroBinarySensorEntityDescription,
        zone_id: int,
    ) -> None:
        """Initialize the Netro sensor."""
        super().__init__(coordinator)
        self.entity_description = description
        self.zone_id = zone_id
        self._attr_unique_id = (
            f"{coordinator.active_zones[zone_id].serial_number}-{description.key}"
        )
        self._attr_device_info = coordinator.active_zones[zone_id].device_info

    @property
    def is_on(self) -> bool | None:
        """Return True if the current zone is currently watering."""
        return getattr(
            self.coordinator.active_zones[self.zone_id],
            self.entity_description.netro_name,
        )

    @property
    def extra_state_attributes(self) -> dict[str, Any]:
        """Return state attributes."""
        zone_attributes = {
            "zone id": self.zone_id,
            "update interval": f"{round(self.coordinator.update_interval.total_seconds() / 60)} mn"
            if self.coordinator.update_interval is not None
            else None,
        }
        if self.coordinator.current_slowdown_factor > 1:
            zone_attributes[
                "slowdown factor"
            ] = self.coordinator.current_slowdown_factor
        meta_attributes = {
            EXTRA_STATE_ATTRIBUTE_SEP_LEFT: EXTRA_STATE_ATTRIBUTE_SEP_RIGHT,
            "request time (UTC)": self.coordinator.metadata.time
            if self.coordinator.metadata is not None
            else None,
            "last active date": dt_util.as_local(
                self.coordinator.metadata.last_active_date.replace(
                    tzinfo=datetime.UTC
                )
            )
            if self.coordinator.metadata is not None
            else None,
            "transaction id": self.coordinator.metadata.tid
            if self.coordinator.metadata is not None
            else None,
            "token limit": self.coordinator.metadata.token_limit
            if self.coordinator.metadata is not None
            else None,
            "token remaining": self.coordinator.metadata.token_remaining
            if self.coordinator.metadata is not None
            else None,
            "token reset": dt_util.as_local(
                self.coordinator.metadata.token_reset_date.replace(
                    tzinfo=datetime.UTC
                )
            )
            if self.coordinator.metadata is not None
            else None,
        }
        return zone_attributes | meta_attributes
