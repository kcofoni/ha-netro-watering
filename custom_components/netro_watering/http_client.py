"""Real HTTP client implementation using aiohttp."""

from __future__ import annotations

import logging
from typing import Any, Self

import aiohttp
from pynetro.client import mask
from pynetro.http import AsyncHTTPResponse

_LOGGER = logging.getLogger(__name__)


class AiohttpResponse:
    """Adapter for aiohttp.ClientResponse to match AsyncHTTPResponse protocol."""

    def __init__(self, response: aiohttp.ClientResponse) -> None:
        """Initialize with aiohttp response."""
        self._response = response
        self.status = response.status

    async def json(self) -> Any:
        """Return JSON data from response, with Netro keys masked for logging."""
        data = await self._response.json()

        # Masque les clÃ©s Netro dans le JSON avant de logger
        def _mask_json(obj):
            if isinstance(obj, dict):
                return {
                    k: mask(v) if k in {"serial", "key"} else _mask_json(v)
                    for k, v in obj.items()
                }
            if isinstance(obj, list):
                return [_mask_json(i) for i in obj]
            return obj

        masked_data = _mask_json(data)
        _LOGGER.debug("HTTP response JSON: %s", masked_data)
        return data

    async def text(self) -> str:
        """Return text data from response."""
        return await self._response.text()

    def raise_for_status(self) -> None:
        """Raise exception for HTTP errors."""
        self._response.raise_for_status()

    async def __aenter__(self) -> Self:
        """Context manager entry."""
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        """Context manager exit."""
        self._response.close()


class AiohttpClient:
    """Real HTTP client using aiohttp that implements AsyncHTTPClient protocol."""

    def __init__(self, session: aiohttp.ClientSession | None = None) -> None:
        """Initialize with optional session."""
        self._session = session
        self._own_session = session is None

    async def __aenter__(self) -> Self:
        """Context manager entry."""
        if self._own_session:
            self._session = aiohttp.ClientSession()
        return self

    async def __aexit__(self, exc_type, exc, tb) -> None:
        """Context manager exit."""
        if self._own_session and self._session:
            await self._session.close()

    def get(self, url: str, **kwargs) -> AsyncHTTPResponse:
        """Make GET request."""
        if not self._session:
            raise RuntimeError(
                "Session not initialized. Use 'async with' context manager."
            )
        log_kwargs = _mask_key_in_kwargs(kwargs)
        _LOGGER.debug("HTTP GET %s params=%s", url, log_kwargs)
        aiohttp_kwargs = self._adapt_kwargs(kwargs)
        response = self._session.get(url, **aiohttp_kwargs)
        return AiohttpResponseContextManager(response)

    def post(self, url: str, **kwargs) -> AsyncHTTPResponse:
        """Make POST request."""
        if not self._session:
            raise RuntimeError(
                "Session not initialized. Use 'async with' context manager."
            )
        log_kwargs = _mask_key_in_kwargs(kwargs)
        _LOGGER.debug("HTTP POST %s params=%s", url, log_kwargs)
        aiohttp_kwargs = self._adapt_kwargs(kwargs)
        response = self._session.post(url, **aiohttp_kwargs)
        return AiohttpResponseContextManager(response)

    def put(self, url: str, **kwargs) -> AsyncHTTPResponse:
        """Make PUT request."""
        if not self._session:
            raise RuntimeError(
                "Session not initialized. Use 'async with' context manager."
            )
        log_kwargs = _mask_key_in_kwargs(kwargs)
        _LOGGER.debug("HTTP PUT %s params=%s", url, log_kwargs)
        aiohttp_kwargs = self._adapt_kwargs(kwargs)
        response = self._session.put(url, **aiohttp_kwargs)
        return AiohttpResponseContextManager(response)

    def delete(self, url: str, **kwargs) -> AsyncHTTPResponse:
        """Make DELETE request."""
        if not self._session:
            raise RuntimeError(
                "Session not initialized. Use 'async with' context manager."
            )
        log_kwargs = _mask_key_in_kwargs(kwargs)
        _LOGGER.debug("HTTP DELETE %s params=%s", url, log_kwargs)
        aiohttp_kwargs = self._adapt_kwargs(kwargs)
        response = self._session.delete(url, **aiohttp_kwargs)
        return AiohttpResponseContextManager(response)

    def _adapt_kwargs(self, kwargs: dict[str, Any]) -> dict[str, Any]:
        """Adapt generic HTTP kwargs to aiohttp format."""
        adapted = {}

        if "headers" in kwargs:
            adapted["headers"] = kwargs["headers"]

        if "params" in kwargs:
            adapted["params"] = kwargs["params"]

        if "timeout" in kwargs:
            # aiohttp utilise aiohttp.ClientTimeout
            adapted["timeout"] = aiohttp.ClientTimeout(total=kwargs["timeout"])

        if "json" in kwargs:
            adapted["json"] = kwargs["json"]

        if "data" in kwargs:
            adapted["data"] = kwargs["data"]

        return adapted


class AiohttpResponseContextManager:
    """Context manager wrapper for aiohttp response."""

    def __init__(self, response_coro) -> None:
        """Initialize with response coroutine."""
        self._response_coro = response_coro
        self._response = None

    async def __aenter__(self) -> AiohttpResponse:
        """Enter context and get response."""
        aiohttp_response = await self._response_coro.__aenter__()
        self._response = AiohttpResponse(aiohttp_response)
        return self._response

    async def __aexit__(self, exc_type, exc, tb) -> None:
        """Exit context."""
        if self._response:
            await self._response.__aexit__(exc_type, exc, tb)
        await self._response_coro.__aexit__(exc_type, exc, tb)


def _mask_key_in_kwargs(kwargs: dict[str, Any]) -> dict[str, Any]:
    """Return a copy of kwargs with Netro keys masked for logging."""
    masked = kwargs.copy()
    if (
        "json" in masked
        and isinstance(masked["json"], dict)
        and "key" in masked["json"]
    ):
        masked["json"] = masked["json"].copy()
        masked["json"]["key"] = mask(masked["json"]["key"])
    if (
        "params" in masked
        and isinstance(masked["params"], dict)
        and "key" in masked["params"]
    ):
        masked["params"] = masked["params"].copy()
        masked["params"]["key"] = mask(masked["params"]["key"])
    return masked
