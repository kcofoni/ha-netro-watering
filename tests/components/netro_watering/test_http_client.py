"""Tests for http_client.py classes and functions."""

from unittest.mock import AsyncMock, MagicMock, patch

import aiohttp
import pytest

from custom_components.netro_watering.http_client import (
    AiohttpClient,
    AiohttpResponse,
    AiohttpResponseContextManager,
    _mask_key_in_kwargs,
)


class TestAiohttpResponse:
    """Test suite for AiohttpResponse class."""

    @pytest.fixture
    def mock_aiohttp_response(self):
        """Create a mock aiohttp.ClientResponse."""
        mock_response = MagicMock(spec=aiohttp.ClientResponse)
        mock_response.status = 200
        mock_response.json = AsyncMock()
        mock_response.text = AsyncMock()
        mock_response.raise_for_status = MagicMock()
        mock_response.close = MagicMock()
        return mock_response

    def test_aiohttpresponse_init(self, mock_aiohttp_response, snapshot):
        """Test AiohttpResponse initialization."""
        response = AiohttpResponse(mock_aiohttp_response)

        result = {
            "status": response.status,
            "has_response": response._response is mock_aiohttp_response,
        }

        assert result == snapshot

    async def test_aiohttpresponse_json_simple_data(
        self, mock_aiohttp_response, snapshot
    ):
        """Test AiohttpResponse.json() with simple data."""
        test_data = {"message": "success", "data": [1, 2, 3]}
        mock_aiohttp_response.json.return_value = test_data

        response = AiohttpResponse(mock_aiohttp_response)

        with patch(
            "custom_components.netro_watering.http_client._LOGGER"
        ) as mock_logger:
            result = await response.json()

            assert result == test_data
            mock_logger.debug.assert_called_once()
            log_call_args = mock_logger.debug.call_args[0]

            result_data = {
                "returned_data": result,
                "log_message_template": log_call_args[0],
                "logged_data": log_call_args[1],
            }

            assert result_data == snapshot

    async def test_aiohttpresponse_json_with_sensitive_keys(
        self, mock_aiohttp_response, snapshot
    ):
        """Test AiohttpResponse.json() masks sensitive Netro keys."""
        test_data = {
            "serial": "ABC123456789",
            "key": "secret_api_key_12345",
            "message": "success",
            "devices": [
                {"serial": "DEF987654321", "name": "Controller1"},
                {"key": "another_secret", "id": 42},
            ],
            "nested": {
                "serial": "nested_serial",
                "key": "nested_key",
                "data": "normal_data",
            },
        }
        mock_aiohttp_response.json.return_value = test_data

        response = AiohttpResponse(mock_aiohttp_response)

        with patch(
            "custom_components.netro_watering.http_client._LOGGER"
        ) as mock_logger:
            result = await response.json()

            # Original data should be returned unchanged
            assert result == test_data

            # But logged data should be masked
            mock_logger.debug.assert_called_once()
            log_call_args = mock_logger.debug.call_args[0]
            logged_data = log_call_args[1]

            result_data = {
                "original_data_unchanged": result == test_data,
                "logged_data": logged_data,
                "sensitive_keys_masked": {
                    "serial_masked": "***" in str(logged_data),
                    "key_masked": "***" in str(logged_data),
                },
            }

            assert result_data == snapshot

    async def test_aiohttpresponse_text(self, mock_aiohttp_response, snapshot):
        """Test AiohttpResponse.text() method."""
        test_text = "This is a text response"
        mock_aiohttp_response.text.return_value = test_text

        response = AiohttpResponse(mock_aiohttp_response)
        result = await response.text()

        assert result == snapshot

    def test_aiohttpresponse_raise_for_status(self, mock_aiohttp_response, snapshot):
        """Test AiohttpResponse.raise_for_status() method."""
        response = AiohttpResponse(mock_aiohttp_response)

        # Should not raise when no exception
        response.raise_for_status()
        mock_aiohttp_response.raise_for_status.assert_called_once()

        # Test when raises exception
        mock_aiohttp_response.raise_for_status.side_effect = (
            aiohttp.ClientResponseError(
                request_info=MagicMock(), history=(), status=404, message="Not Found"
            )
        )

        with pytest.raises(aiohttp.ClientResponseError):
            response.raise_for_status()

        result = {
            "raise_for_status_called": mock_aiohttp_response.raise_for_status.call_count
            == 2,
            "exception_propagated": True,
        }

        assert result == snapshot

    async def test_aiohttpresponse_context_manager(
        self, mock_aiohttp_response, snapshot
    ):
        """Test AiohttpResponse as context manager."""
        response = AiohttpResponse(mock_aiohttp_response)

        async with response as ctx_response:
            assert ctx_response is response

        # Check that close was called on exit
        mock_aiohttp_response.close.assert_called_once()

        result = {
            "context_manager_works": True,
            "close_called_on_exit": mock_aiohttp_response.close.call_count == 1,
        }

        assert result == snapshot


class TestAiohttpClient:
    """Test suite for AiohttpClient class."""

    @pytest.fixture
    def mock_session(self):
        """Create a mock aiohttp.ClientSession."""
        session = AsyncMock(spec=aiohttp.ClientSession)
        session.get = MagicMock()
        session.post = MagicMock()
        session.put = MagicMock()
        session.delete = MagicMock()
        session.close = AsyncMock()
        return session

    def test_aiohttpclient_init_with_session(self, mock_session, snapshot):
        """Test AiohttpClient initialization with provided session."""
        client = AiohttpClient(session=mock_session)

        result = {
            "session_is_provided": client._session is mock_session,
            "own_session_flag": client._own_session,
        }

        assert result == snapshot

    def test_aiohttpclient_init_without_session(self, snapshot):
        """Test AiohttpClient initialization without session."""
        client = AiohttpClient()

        result = {
            "session_is_none": client._session is None,
            "own_session_flag": client._own_session,
        }

        assert result == snapshot

    async def test_aiohttpclient_context_manager_own_session(self, snapshot):
        """Test AiohttpClient as context manager when it creates its own session."""
        client = AiohttpClient()  # No session provided

        with patch("aiohttp.ClientSession") as mock_session_class:
            mock_session = AsyncMock()
            mock_session_class.return_value = mock_session

            async with client as ctx_client:
                assert ctx_client is client
                assert client._session is mock_session

            # Session should be closed on exit
            mock_session.close.assert_called_once()

        result = {
            "context_manager_works": True,
            "session_created": True,
            "session_closed": True,
        }

        assert result == snapshot

    async def test_aiohttpclient_context_manager_provided_session(
        self, mock_session, snapshot
    ):
        """Test AiohttpClient as context manager with provided session."""
        client = AiohttpClient(session=mock_session)

        async with client as ctx_client:
            assert ctx_client is client
            assert client._session is mock_session

        # Provided session should NOT be closed
        mock_session.close.assert_not_called()

        result = {
            "context_manager_works": True,
            "provided_session_not_closed": mock_session.close.call_count == 0,
        }

        assert result == snapshot

    def test_aiohttpclient_get_without_session(self, snapshot):
        """Test AiohttpClient.get() without initialized session."""
        client = AiohttpClient()

        with pytest.raises(RuntimeError) as exc_info:
            client.get("http://example.com")

        result = {
            "error_message": str(exc_info.value),
            "error_type": type(exc_info.value).__name__,
        }

        assert result == snapshot

    def test_aiohttpclient_get_with_session(self, mock_session, snapshot):
        """Test AiohttpClient.get() with session."""
        client = AiohttpClient(session=mock_session)

        # Mock the session.get to return a coroutine
        mock_coro = AsyncMock()
        mock_session.get.return_value = mock_coro

        with patch(
            "custom_components.netro_watering.http_client._LOGGER"
        ) as mock_logger:
            result = client.get("http://example.com", params={"test": "value"})

            # Should return AiohttpResponseContextManager
            assert isinstance(result, AiohttpResponseContextManager)

            # Session.get should be called with adapted kwargs
            mock_session.get.assert_called_once()
            call_args = mock_session.get.call_args

            result_data = {
                "url": call_args[0][0],
                "kwargs_keys": list(call_args[1].keys()),
                "logger_called": mock_logger.debug.called,
            }

            assert result_data == snapshot

    def test_aiohttpclient_post_with_json(self, mock_session, snapshot):
        """Test AiohttpClient.post() with JSON data."""
        client = AiohttpClient(session=mock_session)

        mock_coro = AsyncMock()
        mock_session.post.return_value = mock_coro

        test_data = {"key": "secret123", "data": "test"}

        with patch(
            "custom_components.netro_watering.http_client._LOGGER"
        ) as mock_logger:
            result = client.post("http://example.com", json=test_data)

            assert isinstance(result, AiohttpResponseContextManager)

            # Check that logging masks sensitive data
            mock_logger.debug.assert_called_once()
            log_args = mock_logger.debug.call_args[0]

            result_data = {
                "method_called": "POST",
                "log_contains_masked_key": "***" in str(log_args),
                "session_post_called": mock_session.post.called,
            }

            assert result_data == snapshot

    def test_aiohttpclient_put_method(self, mock_session, snapshot):
        """Test AiohttpClient.put() method."""
        client = AiohttpClient(session=mock_session)

        mock_coro = AsyncMock()
        mock_session.put.return_value = mock_coro

        with patch(
            "custom_components.netro_watering.http_client._LOGGER"
        ) as mock_logger:
            result = client.put("http://example.com", json={"update": "data"})

            assert isinstance(result, AiohttpResponseContextManager)
            mock_session.put.assert_called_once()

            result_data = {
                "method_called": "PUT",
                "session_put_called": mock_session.put.called,
                "logger_called": mock_logger.debug.called,
            }

            assert result_data == snapshot

    def test_aiohttpclient_delete_method(self, mock_session, snapshot):
        """Test AiohttpClient.delete() method."""
        client = AiohttpClient(session=mock_session)

        mock_coro = AsyncMock()
        mock_session.delete.return_value = mock_coro

        with patch(
            "custom_components.netro_watering.http_client._LOGGER"
        ) as mock_logger:
            result = client.delete("http://example.com", params={"id": "123"})

            assert isinstance(result, AiohttpResponseContextManager)
            mock_session.delete.assert_called_once()

            result_data = {
                "method_called": "DELETE",
                "session_delete_called": mock_session.delete.called,
                "logger_called": mock_logger.debug.called,
            }

            assert result_data == snapshot

    def test_aiohttpclient_all_methods_without_session_error(self, snapshot):
        """Test that all HTTP methods raise RuntimeError without session."""
        client = AiohttpClient()

        methods_to_test = [
            ("get", lambda: client.get("http://example.com")),
            ("post", lambda: client.post("http://example.com")),
            ("put", lambda: client.put("http://example.com")),
            ("delete", lambda: client.delete("http://example.com")),
        ]

        results = []
        for method_name, method_call in methods_to_test:
            with pytest.raises(RuntimeError) as exc_info:
                method_call()

            results.append(
                {
                    "method": method_name,
                    "error_message": str(exc_info.value),
                    "error_type": type(exc_info.value).__name__,
                }
            )

        assert results == snapshot

    def test_aiohttpclient_adapt_kwargs(self, mock_session, snapshot):
        """Test AiohttpClient._adapt_kwargs() method."""
        client = AiohttpClient(session=mock_session)

        input_kwargs = {
            "headers": {"Content-Type": "application/json"},
            "params": {"test": "value"},
            "timeout": 30,
            "json": {"data": "test"},
            "data": "raw_data",
            "unknown_param": "should_be_ignored",
        }

        result = client._adapt_kwargs(input_kwargs)

        # Check that timeout is converted to ClientTimeout
        timeout_obj = result.get("timeout")
        has_correct_timeout = (
            timeout_obj is not None
            and hasattr(timeout_obj, "total")
            and timeout_obj.total == 30
        )

        result_data = {
            "headers": result.get("headers"),
            "params": result.get("params"),
            "json": result.get("json"),
            "data": result.get("data"),
            "has_timeout_object": has_correct_timeout,
            "unknown_param_ignored": "unknown_param" not in result,
            "result_keys": sorted(result.keys()),
        }

        assert result_data == snapshot


class TestAiohttpResponseContextManager:
    """Test suite for AiohttpResponseContextManager class."""

    async def test_response_context_manager(self, snapshot):
        """Test AiohttpResponseContextManager functionality."""
        # Create mock coroutine that returns a mock aiohttp response
        mock_aiohttp_response = MagicMock(spec=aiohttp.ClientResponse)
        mock_aiohttp_response.status = 200

        mock_coro = AsyncMock()
        mock_coro.__aenter__ = AsyncMock(return_value=mock_aiohttp_response)
        mock_coro.__aexit__ = AsyncMock()

        ctx_manager = AiohttpResponseContextManager(mock_coro)

        async with ctx_manager as response:
            assert isinstance(response, AiohttpResponse)
            assert response.status == 200
            assert response._response is mock_aiohttp_response

        # Check that both context managers were properly called
        mock_coro.__aenter__.assert_called_once()
        mock_coro.__aexit__.assert_called_once()

        result = {
            "context_manager_works": True,
            "returns_aiohttpresponse": True,
            "status_correct": True,
            "cleanup_called": True,
        }

        assert result == snapshot


class TestMaskKeyInKwargs:
    """Test suite for _mask_key_in_kwargs function."""

    def test_mask_key_in_kwargs_no_sensitive_data(self, snapshot):
        """Test _mask_key_in_kwargs with no sensitive data."""
        kwargs = {
            "headers": {"Content-Type": "application/json"},
            "params": {"test": "value"},
            "data": "some data",
        }

        result = _mask_key_in_kwargs(kwargs)

        # Should return a copy with no changes
        assert result == kwargs
        assert result is not kwargs  # Should be a copy
        assert result == snapshot

    def test_mask_key_in_kwargs_with_json_key(self, snapshot):
        """Test _mask_key_in_kwargs masks key in json data."""
        kwargs = {
            "json": {
                "key": "secret_api_key_12345",
                "data": "normal_data",
                "serial": "not_masked_here",  # Only 'key' is masked in json
            },
            "params": {"test": "value"},
        }

        result = _mask_key_in_kwargs(kwargs)

        # Original should be unchanged
        assert kwargs["json"]["key"] == "secret_api_key_12345"

        # Result should have masked key
        result_data = {
            "original_unchanged": kwargs["json"]["key"] == "secret_api_key_12345",
            "result_key_masked": "***" in result["json"]["key"],
            "other_json_data_unchanged": result["json"]["data"] == "normal_data",
            "serial_not_masked": result["json"]["serial"] == "not_masked_here",
            "params_unchanged": result["params"] == kwargs["params"],
            "is_copy": result is not kwargs,
        }

        assert result_data == snapshot

    def test_mask_key_in_kwargs_with_params_key(self, snapshot):
        """Test _mask_key_in_kwargs masks key in params data."""
        kwargs = {
            "params": {
                "key": "secret_param_key",
                "other": "value",
            },
            "json": {"data": "test"},
        }

        result = _mask_key_in_kwargs(kwargs)

        # Original should be unchanged
        assert kwargs["params"]["key"] == "secret_param_key"

        result_data = {
            "original_unchanged": kwargs["params"]["key"] == "secret_param_key",
            "result_key_masked": "***" in result["params"]["key"],
            "other_params_unchanged": result["params"]["other"] == "value",
            "json_unchanged": result["json"] == kwargs["json"],
            "is_copy": result is not kwargs,
        }

        assert result_data == snapshot

    def test_mask_key_in_kwargs_both_json_and_params(self, snapshot):
        """Test _mask_key_in_kwargs with keys in both json and params."""
        kwargs = {
            "json": {"key": "json_secret_key", "data": "test"},
            "params": {"key": "params_secret_key", "other": "value"},
            "headers": {"Content-Type": "application/json"},
        }

        result = _mask_key_in_kwargs(kwargs)

        result_data = {
            "json_key_masked": "***" in result["json"]["key"],
            "params_key_masked": "***" in result["params"]["key"],
            "json_data_unchanged": result["json"]["data"] == "test",
            "params_other_unchanged": result["params"]["other"] == "value",
            "headers_unchanged": result["headers"] == kwargs["headers"],
            "originals_unchanged": (
                kwargs["json"]["key"] == "json_secret_key"
                and kwargs["params"]["key"] == "params_secret_key"
            ),
        }

        assert result_data == snapshot

    def test_mask_key_in_kwargs_edge_cases(self, snapshot):
        """Test _mask_key_in_kwargs with edge cases."""
        test_cases = [
            # Non-dict json
            {"json": "not_a_dict", "params": {"test": "value"}},
            # Non-dict params
            {"json": {"key": "secret"}, "params": "not_a_dict"},
            # Empty dicts
            {"json": {}, "params": {}},
            # No json or params
            {"headers": {"test": "value"}},
        ]

        results = []
        for i, kwargs in enumerate(test_cases):
            result = _mask_key_in_kwargs(kwargs)
            results.append(
                {
                    "case": i,
                    "input": kwargs,
                    "output": result,
                    "is_copy": result is not kwargs,
                    "unchanged": result == kwargs,
                }
            )

        assert results == snapshot
