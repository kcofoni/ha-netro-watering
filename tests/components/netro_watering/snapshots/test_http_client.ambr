# serializer version: 1
# name: TestAiohttpClient.test_aiohttpclient_adapt_kwargs
  dict({
    'data': 'raw_data',
    'has_timeout_object': True,
    'headers': dict({
      'Content-Type': 'application/json',
    }),
    'json': dict({
      'data': 'test',
    }),
    'params': dict({
      'test': 'value',
    }),
    'result_keys': list([
      'data',
      'headers',
      'json',
      'params',
      'timeout',
    ]),
    'unknown_param_ignored': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_all_methods_without_session_error
  list([
    dict({
      'error_message': "Session not initialized. Use 'async with' context manager.",
      'error_type': 'RuntimeError',
      'method': 'get',
    }),
    dict({
      'error_message': "Session not initialized. Use 'async with' context manager.",
      'error_type': 'RuntimeError',
      'method': 'post',
    }),
    dict({
      'error_message': "Session not initialized. Use 'async with' context manager.",
      'error_type': 'RuntimeError',
      'method': 'put',
    }),
    dict({
      'error_message': "Session not initialized. Use 'async with' context manager.",
      'error_type': 'RuntimeError',
      'method': 'delete',
    }),
  ])
# ---
# name: TestAiohttpClient.test_aiohttpclient_context_manager_own_session
  dict({
    'context_manager_works': True,
    'session_closed': True,
    'session_created': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_context_manager_provided_session
  dict({
    'context_manager_works': True,
    'provided_session_not_closed': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_delete_method
  dict({
    'logger_called': True,
    'method_called': 'DELETE',
    'session_delete_called': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_get_with_session
  dict({
    'kwargs_keys': list([
      'params',
    ]),
    'logger_called': True,
    'url': 'http://example.com',
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_get_without_session
  dict({
    'error_message': "Session not initialized. Use 'async with' context manager.",
    'error_type': 'RuntimeError',
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_init_with_session
  dict({
    'own_session_flag': False,
    'session_is_provided': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_init_without_session
  dict({
    'own_session_flag': True,
    'session_is_none': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_post_with_json
  dict({
    'log_contains_masked_key': True,
    'method_called': 'POST',
    'session_post_called': True,
  })
# ---
# name: TestAiohttpClient.test_aiohttpclient_put_method
  dict({
    'logger_called': True,
    'method_called': 'PUT',
    'session_put_called': True,
  })
# ---
# name: TestAiohttpResponse.test_aiohttpresponse_context_manager
  dict({
    'close_called_on_exit': True,
    'context_manager_works': True,
  })
# ---
# name: TestAiohttpResponse.test_aiohttpresponse_init
  dict({
    'has_response': True,
    'status': 200,
  })
# ---
# name: TestAiohttpResponse.test_aiohttpresponse_json_simple_data
  dict({
    'log_message_template': 'HTTP response JSON: %s',
    'logged_data': dict({
      'data': list([
        1,
        2,
        3,
      ]),
      'message': 'success',
    }),
    'returned_data': dict({
      'data': list([
        1,
        2,
        3,
      ]),
      'message': 'success',
    }),
  })
# ---
# name: TestAiohttpResponse.test_aiohttpresponse_json_with_sensitive_keys
  dict({
    'logged_data': dict({
      'devices': list([
        dict({
          'name': 'Controller1',
          'serial': 'DE********21',
        }),
        dict({
          'id': 42,
          'key': 'an********et',
        }),
      ]),
      'key': 'se********45',
      'message': 'success',
      'nested': dict({
        'data': 'normal_data',
        'key': 'ne********ey',
        'serial': 'ne********al',
      }),
      'serial': 'AB********89',
    }),
    'original_data_unchanged': True,
    'sensitive_keys_masked': dict({
      'key_masked': True,
      'serial_masked': True,
    }),
  })
# ---
# name: TestAiohttpResponse.test_aiohttpresponse_raise_for_status
  dict({
    'exception_propagated': True,
    'raise_for_status_called': True,
  })
# ---
# name: TestAiohttpResponse.test_aiohttpresponse_text
  'This is a text response'
# ---
# name: TestAiohttpResponseContextManager.test_response_context_manager
  dict({
    'cleanup_called': True,
    'context_manager_works': True,
    'returns_aiohttpresponse': True,
    'status_correct': True,
  })
# ---
# name: TestMaskKeyInKwargs.test_mask_key_in_kwargs_both_json_and_params
  dict({
    'headers_unchanged': True,
    'json_data_unchanged': True,
    'json_key_masked': True,
    'originals_unchanged': True,
    'params_key_masked': True,
    'params_other_unchanged': True,
  })
# ---
# name: TestMaskKeyInKwargs.test_mask_key_in_kwargs_edge_cases
  list([
    dict({
      'case': 0,
      'input': dict({
        'json': 'not_a_dict',
        'params': dict({
          'test': 'value',
        }),
      }),
      'is_copy': True,
      'output': dict({
        'json': 'not_a_dict',
        'params': dict({
          'test': 'value',
        }),
      }),
      'unchanged': True,
    }),
    dict({
      'case': 1,
      'input': dict({
        'json': dict({
          'key': 'secret',
        }),
        'params': 'not_a_dict',
      }),
      'is_copy': True,
      'output': dict({
        'json': dict({
          'key': 'se********et',
        }),
        'params': 'not_a_dict',
      }),
      'unchanged': False,
    }),
    dict({
      'case': 2,
      'input': dict({
        'json': dict({
        }),
        'params': dict({
        }),
      }),
      'is_copy': True,
      'output': dict({
        'json': dict({
        }),
        'params': dict({
        }),
      }),
      'unchanged': True,
    }),
    dict({
      'case': 3,
      'input': dict({
        'headers': dict({
          'test': 'value',
        }),
      }),
      'is_copy': True,
      'output': dict({
        'headers': dict({
          'test': 'value',
        }),
      }),
      'unchanged': True,
    }),
  ])
# ---
# name: TestMaskKeyInKwargs.test_mask_key_in_kwargs_no_sensitive_data
  dict({
    'data': 'some data',
    'headers': dict({
      'Content-Type': 'application/json',
    }),
    'params': dict({
      'test': 'value',
    }),
  })
# ---
# name: TestMaskKeyInKwargs.test_mask_key_in_kwargs_with_json_key
  dict({
    'is_copy': True,
    'original_unchanged': True,
    'other_json_data_unchanged': True,
    'params_unchanged': True,
    'result_key_masked': True,
    'serial_not_masked': True,
  })
# ---
# name: TestMaskKeyInKwargs.test_mask_key_in_kwargs_with_params_key
  dict({
    'is_copy': True,
    'json_unchanged': True,
    'original_unchanged': True,
    'other_params_unchanged': True,
    'result_key_masked': True,
  })
# ---
